manim/scene/scene.py

```diff
@@ -915,13 +927,100 @@ def play_internal(self, skip_rendering=False):
+        for method in ("play", "wait", "add", "remove"):
+            self.saved_methods[method] = getattr(self, method)
+            embedded_method = get_embedded_method(method)
+            # Allow for calling scene methods without prepending 'self.'.
+            local_namespace[method] = embedded_method
+            setattr(self, method, embedded_method)
+
+        from IPython.terminal.embed import InteractiveShellEmbed
+        from traitlets.config import Config
+
+        cfg = Config()
+        cfg.TerminalInteractiveShell.confirm_exit = False
+        shell = InteractiveShellEmbed(config=cfg)
+
+        keyboard_thread = threading.Thread(
+            target=ipython,
+            args=(shell, local_namespace),
+        )
+        keyboard_thread.start()
+        self.interact(shell, keyboard_thread)
+
+    def interact(self, shell=None, keyboard_thread=None):
         self.quit_interaction = False
+        keyboard_thread_needs_join = True
         while not (self.renderer.window.is_closing or self.quit_interaction):
-            self.renderer.animation_start_time = 0
-            dt = 1 / config["frame_rate"]
-            self.renderer.render(self, dt, self.moving_mobjects)
-            self.update_mobjects(dt)
+            if not self.queue.empty():
+                tup = self.queue.get_nowait()
+                if tup[0].startswith("rerun"):
+                    kwargs = tup[2]
+                    if "from_animation_number" in kwargs:
+                        config["from_animation_number"] = kwargs[
+                            "from_animation_number"
+                        ]
+                    # # TODO: This option only makes sense if embed_2() is run at the
+                    # # end of a scene by default.
+                    # if "upto_animation_number" in kwargs:
+                    #     config["upto_animation_number"] = kwargs[
+                    #         "upto_animation_number"
+                    #     ]
+
+                    keyboard_thread.join()
+                    raise RerunSceneException
+                elif tup[0].startswith("exit"):
+                    keyboard_thread.join()
+                    keyboard_thread_needs_join = False
+                    break
+                else:
+                    method, args, kwargs = tup
+                    self.saved_methods[method](*args, **kwargs)
+            else:
+                self.renderer.animation_start_time = 0
+                dt = 1 / config["frame_rate"]
+                self.renderer.render(self, dt, self.moving_mobjects)
+                self.update_mobjects(dt)
+
+        # Restore overridden methods.
+        for name, method in self.saved_methods.items():
+            setattr(self, name, method)
+        self.saved_methods = {}
+
+        # Join the keyboard thread if necessary.
+        if shell is not None and keyboard_thread_needs_join:
+            shell.pt_app.app.exit(exception=EOFError)
+            keyboard_thread.join()
+
```

Comment:
To follow best practices, I believe that the logic of `interact` alone, without the logic needed for `embed`, should be in its own function because none of the normal `interact` logic depends on the logic needed for `embed`. As a result, there should be two `interact` functions, one to be used alone and the other is a private function to be used with `interactive_embed`

Reply:
I think the code duplication that this would cause would be worse for ensuring consistency between the two versions.